inc/fftw.hpp

```diff
@@ -0,0 +1,629 @@
+
+      void create() {
+	//TODO: set number of threads
+      
+      }
+
+      void destroy() {
+      
+	//not needed
+      }
+    };
+
+
+    /**
+     * Fftw plan and execution class.
+     *
+     * This class handles:
+     * - {1D, 2D, 3D} x {R2C, C2R, C2C} x {inplace, outplace} x {float, double}.
+     */
+    template<typename TFFT, // see fft_abstract.hpp (FFT_Inplace_Real, ...)
+	     typename TPrecision, // double, float
+	     size_t   NDim // 1..3
+	     >
+    struct FftwImpl
+    {
+
+      //////////////////////////////////////////////////////////////////////////////////////
+      // COMPILE TIME FIELDS
+    
+      using Extent = std::array<std::size_t,(int)NDim>;
+      using Api  = typename traits::plan<TPrecision>;
+      using ComplexType = typename traits::plan<TPrecision>::ComplexType;
+      using RealType = typename traits::plan<TPrecision>::RealType;
+      using PlanType = typename traits::plan<TPrecision>::PlanType;
+    
+      // using FFTExecuteForward  = typename traits::plan::FFTExecuteForward;
+      // using FFTExecuteBackward = typename traits::plan::FFTExecuteBackward;
+
+      static constexpr
+      bool IsInplace = TFFT::IsInplace;
+      static constexpr
+      bool IsComplex = TFFT::IsComplex;
+      static constexpr
+      bool Padding = IsInplace && IsComplex==false && NDim>1;
+
+      using value_type  = typename std::conditional<IsComplex,ComplexType,RealType>::type;
+    
+      // static constexpr
+      //  fftwType FFTForward  = IsComplex ? traits::plan::FFTComplex : traits::plan::FFTForward;
+      // static constexpr
+      //  fftwType FFTBackward = IsComplex ? traits::plan::FFTComplex : traits::plan::FFTBackward;
+
+      //////////////////////////////////////////////////////////////////////////////////////
+      // RUNTIME TIME FIELDS
+    
+      size_t n_;        // =[1]*..*[dim]
+      size_t n_padded_; // =[1]*..*[dim-1]*([dim]/2+1)
+      Extent extents_;
+
+      PlanType			plan_           ;
+      value_type*		data_           = nullptr;
+      ComplexType*       	data_transform_ = nullptr; // intermediate buffer
+      size_t             	data_size_;
+      size_t             	data_transform_size_;
+
+      FftwImpl(const Extent& cextents)
+	: n_(0),
+	  n_padded_(0),
+	  extents_(),
+	  plan_(),
+	  data_(nullptr),
+	  data_transform_(nullptr),
+	  data_size_(0),
+	  data_transform_size_(0)
+      {
+	extents_ = interpret_as::column_major(cextents);
+	
+	n_ = std::accumulate(extents_.begin(),
+			     extents_.end(),
+			     1,
+			     std::multiplies<unsigned>());
+	if(Padding)
+	  n_padded_ = n_ / extents_[NDim-1] * (extents_[NDim-1]/2 + 1);
+
+	data_size_ = ( Padding ? 2*n_padded_ : n_ ) * sizeof(value_type);
+	data_transform_size_ = IsInplace ? 0 : n_ * sizeof(ComplexType);
+
+	traits::thread_api<TPrecision>::init_threads();
+	traits::thread_api<TPrecision>::plan_with_threads(-1);
+      }
+
+      ~FftwImpl(){
+	traits::thread_api<TPrecision>::cleanup_threads();
+      }
```

Comment:
can you move these thread setup and cleanup to Context as you already mentioned in ToDo there?
fftw thread environment will be set up before first benchmark and cleaned up after last benchmark (Context is instantiated gearshifft life-time).